src/collections.h

```diff
@@ -2,580 +2,581 @@
-};
-
-
-template<typename T>
-void
-Collection<T>::pop_head() {
-    // TODO: add emptiness checking
-    Data.erase(Data.begin());
-};
-
-
-template<typename T>
-Collection<T>
-Collection<T>::filter(std::function<bool(T)> func) {
-    std::vector<T> list;
-    for (auto i : Data)
-        if (func(i))
-            list.push_back(i);
-    return Collection<T>(list);
-};
-
-
-template<typename T>
-Collection<T>
-Collection<T>::slice(int low, int high) {
-    std::vector<T> list(high-low);
-    for (int i = 0; i < high-low; i++)
-        list[i] = Data[i+low];
-    return Collection<T>(list);
-};
-
-
-template<typename T>
-Collection<T>&
-Collection<T>::range(int size) {
-    std::vector<T> list(size);
-    for (int i = 0; i < size; i++)
-        list[i] = T(i);
-    static auto x = Collection<T>(list);
-    return x;
-};
-
-
-template<typename T>
-Collection<T>&
-Collection<T>::range(int low, int high) {
-    std::vector<T> list(high-low);
-    for (int i = 0; i < high-low; i++)
-        list[i] = T(low + i);
-    static auto x = Collection<T>(list);
-    return x;
-};
-
-
-template<typename T>
-template<typename Function>
-Collection<typename std::result_of<Function(T)>::type>
-Collection<T>::map(Function func) {
-    using return_type = typename std::result_of<Function(T)>::type;
-
-    std::vector<return_type> list(Data.size());
-    for (int i = 0; i < Data.size(); i++)
-        list[i] = func(Data[i]);
-    return Collection<return_type>(list);
-};
-
-
-template<typename T, typename Function>
-void *
-pmap_thread(void *arg) {
-    struct thread_data {
-        std::vector<T> *list;
-        Function func;
-        int begin;
-        int end;
-    };
-    thread_data *data = ((struct thread_data *)arg);
-    for (int i = data->begin; i < data->end && i < (*(data->list)).size(); i++)
-        (*(data->list))[i] = (data->func)((*(data->list))[i]);
-    pthread_exit(NULL);
-};
-
-
-template<typename T>
-template<typename Function>
-Collection<typename std::result_of<Function(T)>::type>
-Collection<T>::pmap(Function func, int threads) {
-    struct thread_data {
-        std::vector<T> *list;
-        Function func;
-        int begin;
-        int end;
-    };
-    std::vector<pthread_t> thread_pool(threads);
-    std::vector<thread_data> thread_data_pool;
-
-    // TODO: add bounds checking
-    int chunk = Data.size() / threads;
-    int extra = Data.size() - chunk*threads;
-    std::vector<int> indices(extra, chunk+1);
-    std::vector<int> normal(threads-extra, chunk);
-    indices.insert(indices.end(), normal.begin(), normal.end());
-
-    int start = 0;
-    for (int i = 0; i < threads; i++) {
-        pthread_t pid;
-        thread_pool[i] = pid;
-
-        int end = start + indices[i];
-        thread_data td = { &Data, func, start, end};
-        thread_data_pool.push_back(td);
-        start = end;
+    // Return Collection of numeric types over the range [0, size)
+    template<typename T>
+    Collection<T>&
+    range(T size) {
```

Comment:
So if range takes an int, how can we allow users to create ranges of other arithmetic types (double, float, etc)? The only other way I can think of (without using some kind of macro) is giving the function itself a template, so for a range of floats you would write range<float>(1,11). 

But that seems ugly to me